feature: add TypeScript types to the blocks package

johnhooks opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Add TypeScript types to `@wordpress/blocks`

- Part of https://github.com/WordPress/gutenberg/issues/67691.
- Supersedes and closes #43686.

This PR is build on the work already done on [`@types/wordpress__blocks`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/wordpress__blocks) by the following:

Jon Surrell <https://github.com/sirreal>
Dennis Snell <https://github.com/dmsnell>
Tomasz Tunik <https://github.com/tomasztunik>
Lucio Giannotta <https://github.com/sunyatasattva>
Bas Tolen <https://github.com/bastolen>

## Why?

In the conversion regarding @ryanwelcher's PR https://github.com/WordPress/gutenberg/pull/43686, there seemed some consense that having the type definitions separate from the source is brittle. I agree and want to help work toward the solving the issue.

## How?

This PR adds types to `@wordpress/blocks` by defining them in TypeScript files and  applying them to functions using the JSDoc syntax. The end result is to build type definitions directly from the source code.

## Testing Instructions

This PR makes no edits to working code. The type checker and the existing test should provided adequate testing, I think? Opinions on how to further test the types are very welcome. 

## Notes

This is what I consider a first step towards converting the package to TypeScript. At the moment this does not type check the JavaScript files of the package. In this form the types are more helpful to develops of packages that rely on `@wordpress/blocks` than the package itself. Though I would argue having JSDocs pointing to documented types is also a big step in the right direction.

gziolo commented:
Thank you for starting to work on it!

>  It is currently a very big PR, any advice of how best to break it into smaller chucks would be great.

@adamziel or @dmsnell might have good insights on splitting the effort into smaller PRs. They gradually added typings to other big packages, `@wordpress/data` and `@wordpress/core-data` a few months ago.

johnhooks commented:
Thank you, I've been studying the data package. Actually one of the biggest questions I have is about how to correctly type `dispatch` and `select`. I have typed all the actions and selectors, but both `dispatch` and `select` are typed to return an `Object`. An explanation or just pointing me to an example would help a lot.

dmsnell commented:
cc: @sirreal 

I've generally fell into a stance against major changes like this to TypeScript because of some of the fundamental challenges, as it seems you found with `dispatch` and `select`.

IMO a more achievable approach is flipping all of the source files in a package to TS and starting out _extremely permissive_ (even with abundant `any`) and then slowly locking down the types from the leaf nodes. There's been an unfortunate saga of the types being written optimistically only to reject working code because it wasn't obvious when writing the types that other properties were possible.

My assessment for this mismatch between the JS code as written and the TS types is that we tend to have the actual types for an object distributed through seemingly random parts of the codebase. A property may not have any mention except for in a file deep in some seemingly-unrelated file in a different part of the repo and yet application developers rely on that. It's extremely hard to find all these places _and then_ keep the types updated as people are rapidly extending the interfaces in other unrelated development.

As @gziolo noted I spent a lot of time and labor with @adamziel and @sarayourfriend trying to figure out how to add the benefit of the types without blocking legitimate development and it felt near impossible to do. the things that avoided getting in the way tended to be somewhat useless from a type standpoint because we'd have types like `type Thing = { staticProperties } & Record<string, any>`, which is hardly better than `any` or `object`.

But that's kind of what we have to do in the project to avoid being an obstacle for development as long as the TypeScript code interfaces with JS code. Without being 100% TS we lack the tools to uncover all the indirect places that the core code is modified. Even then, we have the further problem that some of the important code is _outside_ the repo, found in plugin code.

---

There was an idea after WCEU last year to try a different approach, which was to analyze things like the selector and actions files and programmatically generate types. While `select` was principally difficult to type in TypeScript, it's not nearly as hard to generate types that would provide value without being part of the code itself, without going through the same type-checking that TS provides. It's a tradeoff that seems more viable given the way most of the code in Gutenberg developed. We diverted to other priorities and never moved forward, but I have some prototype code that uses the TypeScript `npm` library to analyze code and create those type definitions.

johnhooks commented:
> the things that avoided getting in the way tended to be somewhat useless from a type standpoint because we'd have types like type Thing = { staticProperties } & Record<string, any>, which is hardly better than any or object.

@dmsnell I see exactly what you mean. I agree that adding `& Record<string, any>` feels almost unavoidable in a lot of instances which negates a lot of the benefits that would be gained. Type generation could be an answer and I would definitely be willing to explore that.

> A property may not have any mention except for in a file deep in some seemingly-unrelated file in a different part of the repo and yet application developers rely on that.

Some code outside the `blocks` package might use it outside the API and rely on properties that aren't even used in `blocks`?

> I've generally fell into a stance against major changes like this to TypeScript because of some of the fundamental challenges

I understand. Honestly my biggest goal is to learn Gutenberg to the core. I have typically done so in previous projects by exploring the code and this is taking me deep inside how the Blocks work. Even if the entire package isn't updated to TypeScript, the types being completed in [package/blocks/src/types/index.ts](https://github.com/WordPress/gutenberg/blob/6a82d787da97241568dc67df57a76a50999730e9/packages/blocks/src/types/index.ts) could be helpful and at least complete when finished. Do you think there would be benefit to that? And updates to the JSDoc comments?

Oh also another big goal was to start finding out who the TypeScript people are 😄. It seemed like it was going to be hard to do in Slack.

johnhooks commented:
Also I'm totally willing to take what I discover and update the `@types/wordpress__blocks`, if the struggle to include them in the package is to great. It's just not as good as a solution for maintaining the types long term.

dmsnell commented:
@johnhooks I definitely think that starting at the leaf-nodes, so to speak, is the more _achievable_ approach. that is, for example, typing individual selectors with JSDoc types.

a whole-hearted switch from `.js` to `.ts` _everywhere_ would be a big boon, but also presents its own problems, so adding types where we _can_ without introducing architectural changes or comprehensive types is a good way to build gradual typing help. it's something that needs broad agreement and has been a rough discussion so far. maybe you can help!

> Some code outside the blocks package might use it outside the API and rely on properties that aren't even used in blocks?

A good example might be `blockType.supports.__experimentalSelector`. It's not declared anywhere, but some code in the `block-editor` package checks for its presence, and some blocks in the `block-library` happen to include it. The `WPBlockType` type in JSDoc is defined inside of the `blocks` package, where there is no mention of the property. So the risk is that someone looks at that type in the package it comes from and doesn't include the experimental property, and then TypeScript rejects code that adds or relies upon that parameter when in fact it's the type that's wrong, not the code using the property.

johnhooks committed:
build: add tsconfig to package blocks

johnhooks commented:
UPDATE: I am walking back from an attempt of full conversion to TypeScript after some consideration and advice. I am working to integrate types similar to `data` and `core-data`.

johnhooks commented:
There are still `TODO` comments, I'll keep researching them, but right now if a `TODO` is next to a type it typed to be super permissive because there wasn't much information about what it is near its usage. Other `TODO`s note where documentation is missing.

gziolo commented:
> UPDATE: I am walking back from an attempt of full conversion to TypeScript after some consideration and advice. I am working to integrate types similar to `data` and `core-data`.

It sounds like a great idea. Another package that is being gradually migrated to TypeScript is `@wordpress/components`.

> A good example might be `blockType.supports.__experimentalSelector`. It's not declared anywhere, but some code in the `block-editor` package checks for its presence, and some blocks in the `block-library` happen to include it. The `WPBlockType` type in JSDoc is defined inside of the `blocks` package, where there is no mention of the property. So the risk is that someone looks at that type in the package it comes from and doesn't include the experimental property, and then TypeScript rejects code that adds or relies upon that parameter when in fact it's the type that's wrong, not the code using the property.

I got intrigued by how we validate block metadata defined in `block.json` and passed to `registerBlockType` call from `@wordpress/blocks`. I checked the [block.json schema](https://github.com/WordPress/gutenberg/blob/trunk/schemas/json/block.json) maintained in the project, and it avoids mentioning those experimental flags that @dmsnell pointed out. It also doesn't complain about custom properties used when running validation for core blocks on CI:

https://github.com/WordPress/gutenberg/blob/6be1e861fbf7049935a3733a1d2061d360c079f9/test/integration/blocks-schema.test.js#L38-L51

adamziel commented on packages/blocks/README.md:
```diff
@@ -190,9 +190,9 @@ optionally merging new attributes and/or replacing its inner blocks.
--   _newInnerBlocks_ `?Array`: Nested blocks.
+-   _block_ `WPBlock`: Block instance.
+-   _mergeAttributes_ `WPBlockAttributes`: Block attributes.
+-   _newInnerBlocks_ `[WPBlock[]]`: Nested blocks.
```
Is it an array of arrays of WPBlock objects? Or just an array of WPBlock objects? Ditto in line 209

johnhooks replied:
```
/**
 * Returns a block object given its type and attributes.
 *
 * @param {string}            name          Block name.
 * @param {WPBlockAttributes} attributes    Block attributes.
 * @param {WPBlock[]}         [innerBlocks] Nested blocks.
 *
 * @return {WPBlock} Block object.
 */
```

I think something odd is going on in the document build process, because wrapping a parameter name should mean that it is optional. As you see in the JSDoc declaration it its an array of WPBlock that is an optional parameter.

johnhooks replied:
The old JSDocs used the syntax `?Type` but the type checker marks that as `Type|null` which isn't the same as `Type|undefined`. Suggestions on how best to mark optional parameters is welcome. I'm more experienced with TypeScript declarations and JSDocs types are a bit weird.

sirreal replied:
> The old JSDocs used the syntax ?Type but the type checker marks that as Type|null which isn't the same as Type|undefined.

Yup, that was wrong, good spot.

Is this readme generated? If it's generating `[WPBlock[]]` that's a bug in whatever is generating it.

johnhooks replied:
@sirreal Yes it was generated.

sirreal replied:
I'm not actually sure what syntax that generated doc is supposed to be. It looks JSDoc-inspired but the optional `[]` go around the name, not around the type.

[There's an alternate syntax we could try](https://jsdoc.app/tags-param.html#optional-parameters-and-default-values) that should look like this:

```js
/**
 * @param {WPBlock[]=}         innerBlocks Nested blocks.
 */
```

I think it's the [docgen](https://github.com/WordPress/gutenberg/tree/trunk/packages/docgen) package generating these, so we could dig around there and see if that can be improved. We might just need to prefix the description with `Optional` or something :confused:

johnhooks replied:
@sirreal That syntax did not fix the issue.

johnhooks replied:
The `Type=` syntax worked. It no longer looks like a tuple with a single element.

adamziel commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,976 @@
+
+export type WPBlockVariationScope = 'block' | 'inserter' | 'transform';
+
+export interface WPBlockVariation<
```
I noticed some types are prefixed with `WP` and some aren't. Is there a need for this prefix? It's quite clear this is a part of WordPress.

johnhooks replied:
I initially created the types without prefix (as done in `@types/wordpress__blocks`), but then found out that other package mention them with prefix (not imported as a type but in simple JSDocs) and renaming the core types was beyond the scope of what I am trying to accomplish.

I would also prefer un-prefixed, name collision isn't a concern in modules. But I figured it wasn't my decision to make. I agree they should be consistent, if there is agreement on a pattern to follow I can make the changes.

sirreal replied:
Strong vote in favor of moving away from the `WP` prefix. It doesn't add anything.

johnhooks replied:
I removed the `WP` prefix from the types of this packages. We will need to gage the impact this has throughout the documentation. Nothing should break, `@wordpress/blocks` didn't export types before, so no packages could have relied on them. But there is documentation else where in the project that references some of the key interfaces of this package with the `WP` prefix.

dmsnell replied:
we might want to create a discussion on this for the project. I've long been for removing the WP syntax, but some people still preferred them the last time I brought it up.

johnhooks replied:
In the grand scheme of things what they are named isn't the hard part, finding the right definitions and where they are used is difficult. So I'll follow whatever the consensus is, and renaming is easy.

johnhooks replied:
@dmsnell Where is the best place to start a discussion on a types naming convention?

dmsnell replied:
https://github.com/WordPress/gutenberg/discussions/categories/developer-experience?discussions_q=category%3A%22Developer+Experience%22+types

adamziel commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,976 @@
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformShortcode< T extends Record< string, any > > {
```
It would be lovely to make these types exhaustive and don't allow arbitrary properties.

johnhooks replied:
I agree, I can work on that though was warned by @dmsnell that there are concerns about breaking other packages because they might rely on undocumented properties.

johnhooks replied:
I can change it, though I was warned it could prevent acceptance and might break other packages.

adamziel replied:
Yeah, finding all the possible properties is challenging. Exhaustive types or not, this PR is valuable anyway.

adamziel commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,976 @@
+	priority?: number | undefined;
+	blocks: string[];
+	isMatch?( attributes: T, block: string | string[] ): boolean;
+	isMultiBlock?: boolean | undefined;
```
`__experimentalConvert` can also be present in transforms, see `switchToBlockType`()`

johnhooks replied:
I will check that out.

johnhooks replied:
Can it be present in all the different types of transforms?

I'll add a `BaseTransform` that has all the shared properties.

sirreal replied:
On DefinitelyTyped we're staying away from publishing experimental and unstable APIs. I guess to keep the compiler happy here that may not be possible, but maybe we can avoid adding them until they become essential?

Or maybe we can find some way to let folks opt-in or to expose them inside Gutenberg without exposing them to eternal consumers. Maybe some declaration merging tricks where the experimental/unstable types are exposed in a different module? 🤔

johnhooks replied:
@sirreal I have used `@microsoft/api-extractor` before to rollup declarations and it has the ability to set different release levels. It would be a whole other conversation. But there are developed ways to do exactly what you suggest.

adamziel replied:
@sirreal if it's __experimental and merged to WordPress core then it's really public and could use a public type. Here's a related [rabbit hole](https://github.com/WordPress/gutenberg/issues/40316)

adamziel commented:
@johnhooks you've done quite some work here – impressive! I've left a few comments.

As @dmsnell mentioned, it's difficult to identify all the right properties to include in different types. Even more tricky, some objects are populated with data from the server that can be customized by plugins – that probably concerns other packages more than this one but I wanted to make note of it.

Hopefully gradually adding JSDoc types will enable a smoother switch to TS one day. 🙌

johnhooks commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,977 @@
+// Transforms
+// ----------------------------------------------------------------------------
+
+export type BaseTransform = {
+	priority?: number | undefined;
+	__experimentalConvert(
+		blocks: WPBlock | WPBlock[]
+	): WPBlock | WPBlock[] | null;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+};
```
@adamziel Does this look right? 

I just added `BaseTransform` in this commit. Though would `TransformRawSchema` also extend it? It is defined much differently than the others.

adamziel replied:
Looks like BaseTransform is no longer there, I'll resolve this thread

johnhooks commented on packages/blocks/src/api/parser/index.js:
```diff
@@ -21,49 +21,19 @@ import { getBlockAttributes } from './get-block-attributes';
- *
- * @typedef WPRawBlock
- *
- * @property {string=}         blockName    Block name
```
Can we handle instances where a raw block doesn't have a `blockName` in those special cases, rather than have it optional in the defintion? Currently there are many of points `blockName` is accessed without checking for `undefined` so I assume in most cases it is `string`. Opinions or advice?

dmsnell replied:
this occurs when there's freeform HTML within a Gutenberg serialized post. those blocks retain their identity through the editor cycle, so even when re-serializing a post on save they will have no `blockName` (because they aren't inside a block)

if we do this we'll need to discuss making changing and breaking changes to the "non-block content" semanatics, which would have to happen first.

now, with TypeScript we _can_ apply some type magic to remove the optionality of that field, but only in places where we've already confirmed that it's there, such as inside of `block-list/block.js`

johnhooks replied:
@dmsnell I am working hard to only type the existing code during this initial step, not modify how things work. I agree, leaning on TypeScript to remove or add optionality of properties where needed makes the most sense.

johnhooks replied:
I'll start by doing it on a case by case basis and we'll see if any patterns emerge which can be extracted

dmsnell replied:
> not modify how things work

Thank you. I only meant to imply that doing what you were asking would require changing the project, something that's out of scope for a typing PR.

There is opportunity in `block-list/block` to do what you are wanting purely through types, but I don't think it'll be fruitful here for now.

johnhooks replied:
@dmsnell can you add a link to your reference of `block-list/block`?

dmsnell replied:
@johnhooks it's the `packages/block-editor/src/components/block-list/block.js` file

johnhooks commented:
@sirreal and @adamziel I have a question about block attributes sources, what are `raw`, `property`, `node`, and `tag`? Are they internal sources? They aren't listed in [schemas/json/block.json](https://github.com/WordPress/gutenberg/blob/95d52c459cebf7a615c1e442107f9812d50cc7e8/schemas/json/block.json).

https://github.com/WordPress/gutenberg/blob/321313e9089a814e3d3124bd8c49ba3ec447e6fa/packages/blocks/src/api/parser/get-block-attributes.js#L137-L147

sirreal commented:
> @sirreal and @adamziel I have a question about block attributes sources, what are `raw`, `property`, `node`, and `tag`? Are they internal sources?

I'm not sure, I'd have to go poking around source code to figure it out. [They don't seem to be listed in the documentation either.](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#value-source)

(I'm a bit disconnected from the project as of late. Most of my contributions are around TypeScript usage.)

dmsnell commented:
> what are raw, property, node, and tag?

 - `node` and `property` might have been introduced in #33903, but I think were there from long before.
 - `raw` was introduced in #27268 and is used by the HTML block
 - `tag` was introduced in #8767 and is used by the table block

`node` appears to be deprecated and I think both it and `property` are used more in block transforms than in block directly.

johnhooks commented:
@adamziel

There are quite a few issues marked in the code with `TODO` that prevent enabling `checkJs` in the `tsconfig.json`. The issues include dependencies that aren't typed, access of properties defined as optional without type guards, there are also some conflicts between the definitions as they are now and they are used.

I'm thinking to split all these issues into a separate Stage 2 PR, otherwise this one will only grow and get more complicated. If you agree with this plan, I'll reduced this PR to just the type definitions and the JSDoc function signatures. That would unite this package with the types from DefinitelyTyped and remove all that duplication. 

UPDATE: I did to move the TODOs that where in function bodies to another branch.

johnhooks commented on packages/blocks/src/api/node.js:
```diff
@@ -64,9 +58,9 @@ export function getNamedNodeMapAsObject( nodeMap ) {
  * @throws {TypeError} If non-element/text node is passed.
  *
- * @param {Node} domNode DOM node to convert.
+ * @param {Element} domNode DOM node to convert.
```
This parameter and most the other that were originally typed as `Node` accessed the `attributes` property, which isn't present on `Node`, but is on `Element`.

dmsnell replied:
`Node` I believe is the correct type as typed in the JS. we allow `TEXT_NODE` and `ELEMENT_NODE` values for `nodeType`, of which the text node is not an element.

one can argue that in many of these cases we should restrict calling the function on anything else, but if we do that and relax the JS types then we'll end up hitting runtime crashes because of JS code that calls these without the benefit of type-checking.

johnhooks replied:
I am changing them back to `Node`, though it will need more consideration when/if type checking is enabled in this package.

`Property 'querySelector' does not exist on type 'Node'.` in [`matchers.js`](https://github.com/johnhooks/gutenberg/blob/a81221b222b82366fe4578b3d9d6654659f9de76/packages/blocks/src/api/matchers.js#L24)

johnhooks replied:
Also in `hpq` matchers are selectors on `string|Element`.

johnhooks commented on packages/blocks/src/api/registration.js:
```diff
@@ -19,126 +19,32 @@ import { BLOCK_ICON_DEFAULT } from './constants';
-
 function isObject( object ) {
 	return object !== null && typeof object === 'object';
 }
```
Aside: I've noticed guard functions like this through out multiple packages in the monorepo, collecting them into a `@wordpress/common` or `@wordpress/lib` would help simplify the code base and remove duplication.

johnhooks commented on packages/blocks/src/api/registration.js:
```diff
@@ -238,8 +144,9 @@ function getBlockSettingsFromMetadata( { textdomain, ...metadata } ) {
- * @param {Object}        settings            Block settings.
+ * @template {Record<string, any>} Attributes
+ * @param {string|import('../types').BlockConfiguration<Attributes>}                                                        blockNameOrMetadata Block type name or its metadata.
+ * @param {(Partial<import('../types').BlockConfiguration<Attributes>>|import('../types').BlockConfiguration<Attributes>)=} settings            Block settings.
```
I couldn't import the `BlockConfiguration` without supplying the generic argument, so it is imported inline, unlike the other types. If anyone knows a better solution I am open to suggestions.

johnhooks commented on packages/blocks/src/api/serializer.js:
```diff
@@ -365,6 +376,13 @@ export function serializeBlock( block, { isInnerBlocks = false } = {} ) {
 	return getCommentDelimitedContent( blockName, saveAttributes, saveContent );
 }
 
+/**
+ * TODO Undocumented
+ *
+ * @param {Block[]} blocks
+ *
+ * @return {string} TODO Undocumented
+ */
```
I am unsure about how to document this, but something is required in the spaces otherwise there is a `jsdom` ESlint error.

dmsnell replied:
what confusing here? we can add a real description to the unstable function and return type. so far I've generally felt like we should strive to type _what's actually there_ not the ideal interfaces we like to think are there, meaning experimental and unstable APIs need to be in the types

johnhooks replied:
I just was asking for help of how best to document this. I was working to type things and I have been doing my best to find documentation and match it to the types, this is one I had trouble finding information.

johnhooks replied:
And I agree, everything has to be there if it's ever supposed to type check.

johnhooks commented on packages/blocks/src/api/validation/index.js:
```diff
@@ -710,14 +719,16 @@ export function isEquivalentHTML( actual, expected, logger = createLogger() ) {
  */
 
+// TODO ^ What is this?
+
```
This is a duplicate JSDoc comment with a different signature. Is this an overload? Is that possible with JSDocs?

ZebulanStanphill replied:
TypeScript only very recently [added support](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#overload-support-in-jsdoc) for overloads via `@overload`, so it's not that.

It's a typo introduced during a refactor:

https://github.com/WordPress/gutenberg/commit/984160d1368695f22222aea4e20527d7699b842c#diff-ee4697a721079b9d6da3660250b599d45ebdf68c95a055f15f618d38ba73014a

johnhooks replied:
Thanks for the confirmation. I actually just learned about that and did a few `@overloads` on a PR to `hpq` today 😄

johnhooks commented on packages/blocks/src/store/reducer.js:
```diff
@@ -284,6 +295,13 @@ export function categories( state = DEFAULT_CATEGORIES, action ) {
+ * @param {BlockStoreState} state  The current state.
+ * @param {Object}          action Dispatched action.
+ * @return {BlockStoreState} Updated state.
+ */
```
How should collections be documented? What distinguishes it from Categories?

johnhooks commented on packages/core-data/src/entity-provider.js:
```diff
@@ -15,7 +15,7 @@ import { parse, __unstableSerializeAndClean } from '@wordpress/blocks';
 import { STORE_NAME } from './name';
 
-/** @typedef {import('@wordpress/blocks').WPBlock} WPBlock */
+/** @typedef {import('@wordpress/blocks').Block} WPBlock */
```
There was a build error without this import fixed.

johnhooks commented on packages/blocks/src/store/actions.js:
```diff
@@ -47,11 +55,13 @@ function isFunction( maybeFunc ) {
  * Takes the unprocessed block type data and applies all the existing filters for the registered block type.
  * Next, it validates all the settings and performs additional processing to the block type definition.
  *
- * @param {WPBlockType} blockType        Unprocessed block type settings.
- * @param {Object}      thunkArgs        Argument object for the thunk middleware.
- * @param {Function}    thunkArgs.select Function to select from the store.
+ * TODO Originally `thunkArgs.select` was typed as a function, but it is used as an object
  *
- * @return {WPBlockType | undefined} The block, if it has been successfully registered; otherwise `undefined`.
+ * @param {BlockType} blockType        Unprocessed block type settings.
+ * @param {Object}    thunkArgs        Argument object for the thunk middleware.
+ * @param {Object}    thunkArgs.select Function to select from the store.
+ *
+ * @return {BlockType | undefined} The block, if it has been successfully registered; otherwise `undefined`.
```
Originally `thunkArgs.select` was typed as a function, but it is used as an object

johnhooks commented on packages/blocks/src/store/selectors.js:
```diff
@@ -613,11 +622,14 @@ export const getBlockSupport = (
 /**
  * Returns true if the block defines support for a feature, or false otherwise.
  *
- * @param {Object}          state           Data state.
- * @param {(string|Object)} nameOrType      Block name or type object.
- * @param {string}          feature         Feature to test.
- * @param {boolean}         defaultSupports Whether feature is supported by
- *                                          default if not explicitly defined.
+ * TODO The parameter `defaultSupports` is typed `boolean` but shouldn't it be `BlockSupports`?
+ *
+ * @param {BlockStoreState}  state           Data state.
+ * @param {string|BlockType} nameOrType      Block name or type object.
+ * @param {string}           feature         Feature to test.
+ *
+ * @param {boolean}          defaultSupports Whether feature is supported by
+ *                                           default if not explicitly defined.
```
The parameter `defaultSupports` is typed `boolean` but shouldn't it be `BlockSupports`?

dmsnell replied:
what raises this question? if you notice, every instance calling the function passes a boolean value and also the documentation here suggests the same.

johnhooks replied:
I misunderstood of how `getBlockSupports` was typed. I thought its `defaultSupports` parameter was a `BlockSupports` object. I'll remove the TODO and change `getBlockSupports`.

johnhooks reviewed (commented):
I have a few questions that I need a little help on.

dmsnell commented on packages/blocks/src/api/types.ts:
```diff
@@ -0,0 +1,27 @@
+ * 1. `type` (string): Tag name.
+ * 2. `props` (object): Attributes and children array of BlockNode.
+ */
+export type BlockNode = BlockElementNode | string;
```
note that we talk about a "BlockNode" as being the result of the initial parse of a Gutenberg serialized post. this is a block type, set of attributes, inner blocks, and inner HTML.

is that what we're referring to here? It seems like we're talking about something more HTML-specific…

johnhooks replied:
I used the type name used, but not defined, in [node.js](https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/node.js#L26) and [children.js](https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/children.js#L40). Though those are depreciated in favor `html` in [machers.js](https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/matchers.js#L12).

Please advise on a better name for this type.

Edit: I renamed `BlockRaw` to `BlockNode`, then renamed this type you had a question about to `RawBlockNode`, what do you think?

dmsnell replied:
not sure. this is `RichText` stuff isn't it?

`BlockDOMNode`?
Something like that?

dmsnell replied:
cc: @ellatrix @sirreal any thoughts? what _is_ this?

johnhooks replied:
@dmsnell after look at the declarations in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b9748855c21f65681c228df1fb60e72ce489c1d3/types/wordpress__blocks/api/node.d.ts#L36), the parameter `node` (typed `WPBlockNode` in original JSDocs) is typed as `JSX.Element` on DT. Should the `Object` in the JSDoc below  be `JSX.Element`?

https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/node.js#L11-L20

dmsnell replied:
@johnhooks seems to not be a `JSX.Element` - where did you see that?

according to the JSDoc you're sharing, it seems like this is more complicated. it's something like this, but my code here is surely incomplete or inaccurate. `fromDOM` provides the original values of the type

```ts
interface RichTextElement {
    type: string;
    props: Record<string, any>;
}

type RichTextNode = string | RichTextElement;
```

johnhooks replied:
@dmsnell I just found this while reviewing unresolved comments. I matched the types from DT. Take a look at it and see if you think it is correct. The types on DT did not declare a `RichTextElement` type, but typed it as a `JSX.Element`, I will defer to you on what is correct.

johnhooks replied:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__blocks/api/node.d.ts

johnhooks replied:
@dmsnell I'm sorry I missed this comment, I was reviewing all the unresolved comments and found this. I copied the types from [@types/wordpress__blocks](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__blocks/api/node.d.ts). Though They do note match with the original `WPBlockNode` type of the JSDoc

dmsnell commented on packages/blocks/src/api/validation/index.js:
```diff
@@ -21,10 +21,19 @@ import {
 
+/**
+ * @template Value
+ * @param {Value} x The value to identify.
```
note that this isn't about "identifying" but "identity" - it's a function which returns the same value it's given. this is a bit funny, but the foundation of a number of different techniques

johnhooks replied:
I understood the purpose of the function, though I did a bad job describing it. The way its typed is correct? Just the description should be changed?

johnhooks replied:
Is this better?
https://github.com/WordPress/gutenberg/blob/a81221b222b82366fe4578b3d9d6654659f9de76/packages/blocks/src/api/validation/index.js#L32-L37

dmsnell replied:
yes, but we can also add a summary description, even if vague: "Directly returns the value provided."

dmsnell commented on packages/blocks/src/store/reducer.js:
```diff
@@ -36,7 +35,13 @@ export const DEFAULT_CATEGORIES = [
+/**
+ * Key block types by their name.
+ *
+ * @param { BlockType[]} types Block types of which to
```
nitpick: I'm sure you'll find this later, but there's an extra space here

johnhooks replied:
Easy fix, thanks for pointing it out.

dmsnell commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,1504 @@
+ *
+ * @public
+ */
+export type BlockRaw = {
```
this is what we have come to call the "block node," or in a different description, the result of the "stage-1 parse of a block"

johnhooks replied:
If you could adjust the names to fit I would appreciate it. I was doing my best to marry the DefinitelyTyped definitions with what was in the JSDocs, but I'm still getting familiar with everything. That or as a new developer the lifecycle of a block is still a bit of a mystery. I understand it better by looking thought this package, but is there a document that explains the concepts and terms of the block lifecycle?

dmsnell replied:
unfortunately I don't think this is in the code. I started some work #15674 but that's out of date now, and we have held informal conversations around these terms.

the existing references are in this file and we can either choose to keep `RawBlock` (`BlockRaw` is a bit awkward linguistically) or adopt `BlockNode`, which is how I believe we want to refer to this going forward. This term is used in the `children` file.

I realize we talk about `blockNode` in the RichText/DOM stuff, but these are also two different kinds of "node" in play. not sure the best resolution; could be to leave the name clash; could be to pick one and change the other.

johnhooks replied:
I'll take a look at that PR, even if outdated, I'm sure has helpful information.

johnhooks replied:
This is related to https://github.com/WordPress/gutenberg/pull/48604#discussion_r1142826547. I prefer the name `BlockNode` over `BlockRaw` or `RawBlock` for this stage in the block lifecycle.

dmsnell commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,1504 @@
+	 * Inner Blocks.
+	 */
+	innerBlocks: BlockRaw[];
+};
```
I'd be happy to expand on these descriptions if you want

johnhooks replied:
I would greatly appreciate the help

dmsnell replied:
maybe I'll try and post some here in a comment which you can absorb into your branch. if I don't do this by Wednesday please feel free to poke me and remind me to do so 😄

johnhooks replied:
@dmsnell (👈🏻 poke) haha that reminded of poke in old Facebook 😁 Edit: oh apparently it is still a thing

dmsnell replied:
see, I knew it would be good to have you remind me, as I got caught up and didn't get to it… doing this today

dmsnell commented on packages/blocks/src/types.ts:
```diff
@@ -0,0 +1,1504 @@
+/**
+ * @public
+ */
+export namespace AttributeSource {
```
I thought we were wanting to avoid namespaces in TypeScript…

johnhooks replied:
Sure thing, that came over from the DefinitelyTyped declaration. I had added `AttributeSource` to the beginning of everything and it looked pretty verbose, then I moved them to a module, because that is the preferred way to do a namespace, but I wasn't sure what would be preferred so I left it as is.

johnhooks replied:
What would you suggest?

dmsnell replied:
yeah let's leave them in separate files. because of issues with namespaces and JS codebases we don't use them. I constantly forget what those issues are, so I may be wrong here.

johnhooks replied:
I will do that. I was debating on breaking this file up and organizing the types into separate files in a `types` folder. Does that sound like a good idea?

dmsnell replied:
seems like it. can't hurt to try. the blocks package is quite large

johnhooks replied:
I broke everything into separate files. Removes the "namespace", but technically a module namespace now.

dmsnell commented on packages/blocks/src/types/block.ts:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * The block's registered name.
+	 */
+	name: string;
```
```ts
/**
 * Name used to uniquely identify block type, e.g. core/paragraph.
 *
 * The block name comprises a namespace part and a name part separated
 * by a "/". The namespace is usually the name of the plugin providing
 * the block, for example, "myplugin/my-interesting-block".
 *
 * @see https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#block-name
 */
```

johnhooks commented:
Do the playwright tests timeout and fail regularly? They seem to randomly fail and not be related to changes in my commits.

dmsnell commented on packages/blocks/src/types/block.ts:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * Attributes for the block.
+	 */
+	attributes: Attributes;
```
```ts
/**
 * Description of block's attribute types.
 *
 * These define how to source the attributes for a block and
 * may provide default values in case they aren't available.
 *
 * Example
 *     {
 *       "url": {
 *         "type": "string",
 *         "source": "attribute",
 *         "selector": "img",
 *         "attribute": "src"
 *       }
 *     }
 * 
 * @see https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/
 */
```

dmsnell commented on packages/blocks/src/types/block.ts:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * Unique ID registered to the block.
+	 */
+	clientId: string;
```
```ts
/**
 * Auto-generated ID unique to the loaded block inside the editor.
 *
 * Many functions in the editor reference blocks by their client id
 * in order to modify or rearrange blocks. This is set when loading
 * blocks and does not persist between edit sessions.
 *
 * Example:
 *     const clientId = select( blockEditorStore ).getSelectedBlockClientId();
 *     const prevId   = select( blockEditorStore ).getPreviousBlockClientId( clientId );
 *     useDispatch( blockEditorStore ).replaceBlock( prevId, newBlock );
 */
```

dmsnell commented on packages/blocks/src/types/block.ts:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * The parsed HTML content of the block.
+	 */
+	originalContent?: string;
```
```ts
/**
 * Original block HTML from source document, used by the unrecognized block type.
 */
```

dmsnell commented on packages/blocks/src/types/block.ts:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * Validation issues.
+	 */
+	validationIssues?: Array< Record< string, unknown > >; // TODO find `ValidateIssue` type
```
```ts
/**
 * Contains descriptions of any block validation issues that appear in loading the block.
 */
```

dmsnell commented on packages/blocks/src/types/block.ts:
```diff
@@ -0,0 +1,59 @@
+	 *
+	 * @internal
+	 */
+	__unstableBlockSource?: BlockNode;
```
```ts
/**
 * Un-processed original copy of block, used for preserving invalid blocks.
 */
```

johnhooks commented on packages/blocks/src/types/block-node.ts:
```diff
@@ -5,36 +5,49 @@
 
 	/**
-	 * Inner Blocks.
+	 * Nested block nodes; may be empty.
```
@dmsnell I added these descriptions from #48604 still look good?

dmsnell commented:
@johnhooks you also might want to spend some time looking at how things are done in the DefinitelyTyped repo, where a reasonable amount of work has already taken place.

this is not to say we should pull everything from there into Gutenberg, but it may provide some inspiration to you on certain points.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__blocks/index.d.ts#L159

johnhooks commented:
@dmsnell I pulled all the types and interfaces over from DefinitelyTyped and tried to match up documentation that was missing. But I haven't yet gone through the function definitions. I originally typed the functions as I read through the source, I'll do a comparison and fix it up.

dmsnell commented:
> Do the playwright tests timeout and fail regularly? They seem to randomly fail and not be related to changes in my commits.

yes, unfortunately

noahtallen commented on packages/blocks/tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+		{ "path": "../private-apis" }
+	],
+
+	"include": [ "src/**/*", "src/api/i18n-block.json" ]
```
what does adding i18n-block do?

johnhooks replied:
I think this was because the json wasn't included by default and when `checkJs` was on I was getting an error. (I was typically working on this `checkJs` on so I could see the issues.

noahtallen commented:
This approach sounds great to me, and it looks like the typechecker is happy too. I'd recommend rebasing so that #49736 can check the published build types. Sometimes, when `checkJs` is false, we get some errors copied into build-types, and that will make sure we fix them.

What else needs to be done to push this forward? My recent approach has not been to be super careful about the existing DT types. Instead, I've been hoping that we can publish types first, and then gradually improve things as needed. (This way we can stop maintaining DT, which is out of date with parts that don't work correctly anyways, like the store types.)

johnhooks commented:
@noahtallen The biggest decision is to decide how "complete" it needs to be. This diverges from the DT types, and I still need to do a through comparison to the two. Some of it was typed from the inside out, some came from the DT types. I changed some of the type names, I tried to get some guidance from @sirreal, @dmsnell and @adamziel. 

I worked with @aduth to type [`hpq`](https://github.com/aduth/hpq). He and I had plans to improve the types [`memize`](https://github.com/aduth/memize) and [`rememo`](https://github.com/aduth/rememo), but I haven't come back around to it yet. Those are some dependency that would need some work if we were to enable `checkJs`, also my PR #49930 would help appropriately type the block store.

johnhooks commented:
I might need to work on `memize` first.

```
Incorrect published types for /Users/johnhooks/Projects/gutenberg/packages/blocks/build-types/index.d.ts:
 packages/blocks/build-types/api/parser/get-block-attributes.d.ts(95,8): error TS1259: Module '"/Users/johnhooks/Projects/gutenberg/node_modules/memize/index"' can only be default-imported using the 'esModuleInterop' flag

Incorrect published types for /Users/johnhooks/Projects/gutenberg/packages/core-data/build-types/index.d.ts:
 packages/blocks/build-types/api/parser/get-block-attributes.d.ts(95,8): error TS1259: Module '"/Users/johnhooks/Projects/gutenberg/node_modules/memize/index"' can only be default-imported using the 'esModuleInterop' flag
```

noahtallen commented:
Yeah, I had to update `rememo` to get #49736 working!

johnhooks commented:
I have a PR in to memize [#10](https://github.com/aduth/memize/pull/10). It should fix the current issue.

johnhooks commented:
@noahtallen I'm very willing to break this into stages if you think it will help. I hadn't touched this for a month and looking back through, I had two goals. First, integrate the DT types into the package and evolve them, the other was to actually type the internals of the package. The later is a big project and would need actually require modifying working code. I can focus this PR on just providing exported types and extract all the other work for later.

johnhooks commented:
@tyxla this is still failing because I need to get the PR to `memize` through, we have the basics done and tested, it does resolve the issue in this PR, but we need to finalized the actual implementation... so hopefully rather soon. 

Though, I just did a thorough review of this PR and cleaned up a lot of details. 

- There are still a few TODO comments for additional documentation, if someone more familiar with implementation could help with those, it would be great. I did my best to find existing documentation and use it in the newly declared types, but there are some properties or interfaces on which I couldn't find anything.
- I didn't fully type the redux store. The actions would be best to type in TypeScript and rely on the inferred types of the action creators, rather than hard code it. We can work more on this as a follow up after the types exported by `@wordpress/data` are more complete.

tyxla commented:
@johnhooks thanks for the update. I see the PR is still blocked by a new release of `memize`, but let me know when it's ready for another look and I can try to spare a review cycle.

johnhooks commented:
@tyxla other than the `memize` issue I believe this is ready. There are a few `TODO` comments about missing documentation. If you could help fill those in that would be great.

tyxla commented on packages/blocks/src/api/serializer.js:
```diff
@@ -365,6 +376,13 @@ export function serializeBlock( block, { isInnerBlocks = false } = {} ) {
 }
 
+/**
+ * TODO Undocumented
```
```suggestion
 * Serialize and clean blocks to handle empty posts and backward compatibility.
```

tyxla commented on packages/blocks/src/api/types.ts:
```diff
@@ -0,0 +1,55 @@
+ */
+export type BlockSerializationOptions = {
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Whether to output HTML comments around blocks.
```

tyxla commented on packages/blocks/src/api/types.ts:
```diff
@@ -0,0 +1,55 @@
+	isCommentDelimited?: boolean;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Whether we are serializing inner blocks.
```

johnhooks replied:
@tyxla what do you think of this? Is this option only ever used inside the `serializeBlock` function when descending into the inner blocks?

```suggestion
	 * Whether the serialization process has descended into the inner blocks.
```

tyxla replied:
Feel free to go through your suggestion 👍 I think it's more descriptive than what I came up with. In any case, either is much better than having nothing after all.

tyxla commented on packages/blocks/src/store/types.ts:
```diff
@@ -0,0 +1,67 @@
+	groupingBlockName?: string;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Unprocessed block types.
```

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+	| None
+ ) & {
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Role of the block. 
```

tyxla commented on packages/blocks/src/types/block-supports.ts:
```diff
@@ -0,0 +1,372 @@
+	padding: boolean | CSSDirection[];
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Whether to skip the serialization to HTML markup.
	 * Provide `true` to skip serialization for all block supports.
	 * Alternatively, provide an array of block supports for which skip
	 * serialization to HTML markup.
```

tyxla commented on packages/blocks/src/types/block-supports.ts:
```diff
@@ -0,0 +1,372 @@
+	lock?: boolean;
+
+	/**
+	 * A block may want to disable the ability to toggle the lock state.
+	 * It can be locked/unlocked by a user from the block “Options”
+	 * dropdown by default. To disable this behavior, set `lock` to `false`.
```
Just noticed that this seems wrong.

johnhooks replied:
@tyxla totally, it's a copy the the comment above. I'll see if I can find something appropriate for `typography`

tyxla commented on packages/blocks/src/types/block-supports.ts:
```diff
@@ -0,0 +1,372 @@
+	typography?: Partial< TypographyProps >;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * This value signals that a block supports some of the CSS style properties related to shadow.
```

tyxla commented on packages/blocks/src/types/block-supports.ts:
```diff
@@ -0,0 +1,372 @@
+	position: BlockPosition;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * CSS selector to use instead of the default `.wp-block-<blockname>` selector.
```

tyxla commented on packages/blocks/src/types/transform.ts:
```diff
@@ -0,0 +1,162 @@
+	priority?: number;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Whether there are any mobile block transforms.
```

tyxla commented on packages/blocks/src/types/transform.ts:
```diff
@@ -0,0 +1,162 @@
+	usingMobileTransformations?: boolean;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Function that defines how blocks are transformed from particular block types.
```

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+ *
+ * @internal
+ */
+export interface Raw {
+	source: 'raw';
+	type: 'string'; // TODO is this correct?
+}
```
"Raw" seems a bit broad, so I wasn't quite sure what it intends. We should likely prefix it with something, since this type is exported and may mean different things in different contexts.

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+	responsive?: boolean;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Key for a social service (lowercase slug)
```

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+	legacy?: boolean;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Whether the embed provider embeds are responsive.
```

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+	className?: string;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Used by the legacy comments block to specify an instance of the legacy comments block.
```

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+	BlockAttribute< T >
+> & {
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Provide an additional `className`, which used to be supported in the `core/comments-query-loop` block.
```

tyxla commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -0,0 +1,319 @@
+	BlockAttribute< T >
+> & {
+	/**
+	 * TODO Undocumented
```
To all of those descriptions, we could add that they're legacy and are no longer in use.

johnhooks replied:
Do you know what version they were depreciated in?

johnhooks replied:
Would this be enough to add to all of them?
```
@deprecated This is a legacy property and should not be used.
```

tyxla replied:
I'm not particularly sure, we might have to go through them one by one in history. I don't think it's particularly necessary to add the deprecation version to the types.

tyxla commented on packages/blocks/src/types/block-type.ts:
```diff
@@ -0,0 +1,287 @@
+				 */
+				to?: Transform[];
+				/**
+				 * TODO Undocumented
```
```suggestion
				 * The transformations available for mobile devices.
```

tyxla commented on packages/blocks/src/types/block-type.ts:
```diff
@@ -0,0 +1,287 @@
+	): Partial< Attributes >;
+
+	/**
+	 * TODO Undocumented
```
```suggestion
	 * Custom block label callback function
```

tyxla commented on packages/blocks/src/types/block-type.ts:
```diff
@@ -0,0 +1,287 @@
+	): Record< string, string | number | boolean >;
+
+	/**
+	 * TODO Undocumented, but used in core.
```
```suggestion
	 * A function that defines how new attributes are merged with existing ones.
```

tyxla reviewed (commented):
Left some suggestions to fill in the missing documentation. Hope that helps!

johnhooks commented on packages/blocks/src/types/block-attributes.ts:
```diff
@@ -262,39 +262,33 @@ export type BlockAttributes< T extends Record< string, any > = {} > = Record<
 	 * Used in [convert-legacy-block.js](../api/parser/convert-legacy-block.js)
 	 *
-	 * @internal
+	 * @deprecated This is a legacy property and should not be used.
```
@tyxla this look good?

tyxla replied:
Sounds good to me 👍

johnhooks commented:
Update: The changes to `memize`, in v2, solved the previous issue of building type declarations (thank you @aduth). This PR is getting closer and could use a thorough review. There are still a few TODOs regarding missing documentation in JSDoc comments, any further help filling in the holes would be greatly appreciated.

CC: @dmsnell, @gziolo, @noahtallen, and @tyxla

dmsnell commented on packages/blocks/src/store/actions.js:
```diff
@@ -37,7 +45,7 @@ const LEGACY_CATEGORY_MAPPING = {
  *
  * @param {*} maybeFunc The argument to check.
- * @return {boolean} True if the argument is a function, false otherwise.
+ * @return {maybeFunc is Function} True if the argument is a function, false otherwise.
```
nice. first time I've seen this in JSDoc

johnhooks replied:
@dmsnell I'm curious, these sorts of functions seem like they should be in a utils common package, rather than implemented in separate places. I've seen a lot of `isObject` functions stashed in multiple packages of Gutenberg.

dmsnell replied:
these small functions have their own value being spread out:
 - they are allowed to be different as each package has need
 - duplicating them carries little overhead but avoids what could be heavy coupling between packages and what could be pulling in a lot more than someone wants through the dependency

if they were bigger or more complicated it would probably make more sense to build a new package

dmsnell commented:
I will try and get to a more thorough review of this later; for now I need to put it on the TODO list.

gziolo commented:
I wanted to reference a recently opened issue https://github.com/WordPress/gutenberg/issues/53605 where some challenges with typing `registerBlockType` were raised.

johnhooks commented:
@dmsnell @gziolo and @adamziel, is there still interest in this approach? I've been very busy and didn't have time to commit, though I have sometime and willing to resolve the conflicts and continue the review process.

sirreal commented:
@johnhooks Thanks for the update, I'd be very happy to see initiatives to improve TypeScript support and integration like this one move ahead. Thank you!

johnhooks commented on packages/blocks/src/api/registration.js:
```diff
@@ -14,124 +14,34 @@ import { store as blocksStore } from '../store';
- * @property {Component}          edit          Component rendering an element to
- *                                              manipulate the attributes of a block
- *                                              in the context of an editor.
- * @property {WPBlockVariation[]} [variations]  The list of block variations.
```
`variations` is a property I don't have on the `BlockType`.

johnhooks committed:
Apply suggestions from code review

Thanks @tyxla for helping fill in the gaps.

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

johnhooks commented on packages/blocks/src/store/types.ts:
```diff
@@ -0,0 +1,67 @@
+	BlockStyle,
+	BlockType,
+} from '../types';
+
```
@TimothyBJacobs would you recommend explicitly typing out the action types here and using the union of `Action` as the return type of all the action creators?

johnhooks committed:
fix: liniting issues

johnhooks committed:
fix: replace references to WPComponent and WPElement

Use React.ComponentType and React.ReactElement instead of WPComponent and WPElement.

See: [packages/element/src/react.js](https://github.com/WordPress/gutenberg/blob/ca2ee4d6fa8c7044cd1d3045c0985a77659f44fa/packages/element/src/react.js#L37-L47)

noahtallen commented:
Updated my profile so I don't get these anymore :)

manzoorwanijk commented:
Any progress/discussions about this one?

gziolo commented:
@johnhooks, do you plan to continue working on this PR? @manzoorwanijk, how would you like to help move these efforts forward?

manzoorwanijk commented:
> @manzoorwanijk, how would you like to help move these efforts forward?

I am ready to help in whatever way I can.

manzoorwanijk commented:
First things first.

Can we please update this branch from trunk to make it ready for testing?

manzoorwanijk commented on packages/blocks/src/api/parser/index.js:
```diff
@@ -302,10 +274,10 @@ export function parseRawBlock( rawBlock, options ) {
  *
- * @param {string}       content The post content.
- * @param {ParseOptions} options Extra options for handling block parsing.
+ * @param {string}            content The post content.
+ * @param {BlockParseOptions} options Extra options for handling block parsing.
```
`options` is optional

```suggestion
 * @param {string}            content   The post content.
 * @param {BlockParseOptions} [options] Extra options for handling block parsing.
```

gziolo commented:
@manzoorwanijk, this PR is big enough. In the meantime where we wait to hear from @johnhooks. You can also open a few smaller PRs that borrow chunks of code from here.

johnhooks commented:
@gziolo I had determined this wasn't the right approach. Attempting this change in one sweeping PR is too much.

There looks to have been work to add a few types with JSDoc. This PR has some valuable information in the `src/types` directory that could be helpful to continue that effort.

manzoorwanijk commented:
> There looks to have been work to add a few types with JSDoc.

Yes, That's the approach I took in #67416

manzoorwanijk commented:
> You can also open a few smaller PRs that borrow chunks of code from here.

@gziolo I would be happy to make atomic and granular PRs for the conversion provided those are reviewed ASAP to finish the whole process in quick succession.

gziolo commented:
@ryanwelcher opened a PR based on this branch that moves only the types:
- https://github.com/WordPress/gutenberg/pull/48604
